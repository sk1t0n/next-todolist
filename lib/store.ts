import type { User } from "@supabase/supabase-js";
import { create } from "zustand";
import { createClient } from "@/lib/supabase/client";

export interface Todo {
  id: number;
  title: string;
  description: string;
  completed: boolean;
  created_at: string;
  user_id?: string;
}

interface TodoState {
  todos: Todo[];
  newTodo: { title: string; description: string };
  editingTodo: Todo | null;
  isEditing: boolean;
  isLoading: boolean;
  error: string | null;
  user: User | null;

  setTodos: (todos: Todo[]) => void;
  setNewTodo: (todo: { title: string; description: string }) => void;
  setEditingTodo: (todo: Todo | null) => void;
  setIsEditing: (isEditing: boolean) => void;
  setIsLoading: (isLoading: boolean) => void;
  setError: (error: string | null) => void;
  setUser: (user: User | null) => void;

  fetchUser: () => Promise<void>;
  fetchTodos: () => Promise<void>;
  addTodo: () => Promise<void>;
  updateTodo: () => Promise<void>;
  toggleTodo: (id: number, completed: boolean) => Promise<void>;
  deleteTodo: (id: number) => Promise<void>;
}

export const useTodoStore = create<TodoState>((set, get) => ({
  todos: [],
  newTodo: { title: "", description: "" },
  editingTodo: null,
  isEditing: false,
  isLoading: false,
  error: null,
  user: null,

  setTodos: (todos) => set({ todos }),
  setNewTodo: (newTodo) => set({ newTodo }),
  setEditingTodo: (editingTodo) => set({ editingTodo }),
  setIsEditing: (isEditing) => set({ isEditing }),
  setIsLoading: (isLoading) => set({ isLoading }),
  setError: (error) => set({ error }),
  setUser: (user) => set({ user }),

  fetchUser: async () => {
    const supabase = createClient();
    try {
      const {
        data: { user },
      } = await supabase.auth.getUser();
      set({ user });
    } catch {
      set({ user: null });
    }
  },

  fetchTodos: async () => {
    set({ isLoading: true });
    await get().fetchUser();

    const supabase = createClient();
    try {
      const { data, error } = await supabase
        .from("todos")
        .select("*")
        .order("created_at", { ascending: false });

      if (error) {
        set({ error: error.message });
      } else {
        set({ todos: data || [], error: null });
      }
    } catch {
      set({ error: "Failed to fetch todos" });
    } finally {
      set({ isLoading: false });
    }
  },

  addTodo: async () => {
    const { newTodo, user } = get();
    if (!newTodo.title.trim() || !user) return;

    set({ isLoading: true });
    const supabase = createClient();
    try {
      const { data, error } = await supabase
        .from("todos")
        .insert([
          {
            title: newTodo.title,
            description: newTodo.description,
            user_id: user.id,
          },
        ])
        .select();

      if (error) {
        set({ error: error.message });
      } else if (data?.[0]) {
        set((state) => {
          const todos = [...state.todos, data[0]];
          const sortedTodosById = todos.sort((a, b) => b.id - a.id);
          return { todos: sortedTodosById };
        });
        set({ newTodo: { title: "", description: "" }, error: null });
      }
    } catch {
      set({ error: "Failed to add todo" });
    } finally {
      set({ isLoading: false });
    }
  },

  updateTodo: async () => {
    const { editingTodo } = get();
    if (!editingTodo) return;

    set({ isLoading: true });
    const supabase = createClient();
    try {
      const todoForUpdate = {
        title: editingTodo.title,
        description: editingTodo.description,
      };
      const { error } = await supabase
        .from("todos")
        .update(todoForUpdate)
        .eq("id", editingTodo.id);

      if (error) {
        set({ error: error.message });
      } else {
        set((state) => ({
          todos: state.todos.map((todo) =>
            todo.id === editingTodo.id ? { ...todo, ...todoForUpdate } : todo,
          ),
        }));
        set({ editingTodo: null, isEditing: false, error: null });
      }
    } catch {
      set({ error: "Failed to update todo" });
    } finally {
      set({ isLoading: false });
    }
  },

  toggleTodo: async (id: number, completed: boolean) => {
    set({ isLoading: true });
    const supabase = createClient();
    try {
      const todoForUpdate = { completed: !completed };
      const { error } = await supabase
        .from("todos")
        .update(todoForUpdate)
        .eq("id", id);

      if (error) {
        set({ error: error.message });
      } else {
        set((state) => ({
          todos: state.todos.map((todo) =>
            todo.id === id ? { ...todo, ...todoForUpdate } : todo,
          ),
        }));
        set({ error: null });
      }
    } catch {
      set({ error: "Failed to update todo" });
    } finally {
      set({ isLoading: false });
    }
  },

  deleteTodo: async (id: number) => {
    set({ isLoading: true });
    const supabase = createClient();
    try {
      const { error } = await supabase.from("todos").delete().eq("id", id);

      if (error) {
        set({ error: error.message });
      } else {
        set((state) => ({
          todos: state.todos.filter((todo) => todo.id !== id),
        }));
        set({ error: null });
      }
    } catch {
      set({ error: "Failed to delete todo" });
    } finally {
      set({ isLoading: false });
    }
  },
}));
