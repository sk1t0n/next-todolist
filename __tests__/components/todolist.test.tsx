import { render, screen } from "@testing-library/react";
import TodoList from "@/components/todolist";
import { useTodoStore } from "../../lib/store";

jest.mock("../../lib/store", () => {
  const testTodos = [
    {
      id: 1,
      title: "Todo 1",
      description: "",
      completed: false,
      created_at: "2025-01-01T00:00:00Z",
    },
  ];

  let user: any = null;
  const mock = {
    useTodoStore: () => ({
      todos: testTodos,
      newTodo: { title: "", description: "" },
      user,
      fetchUser: () => {},
      fetchTodos: () => {
        return testTodos;
      },
      setUser(newUser: any) {
        user = newUser;
      },
    }),
  };

  return mock;
});

describe("TodoList", () => {
  it("check snapshot", () => {
    const { asFragment } = render(<TodoList />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("show text and todos for guest", () => {
    render(<TodoList />);

    expect(screen.getByText("Todo list")).toBeInTheDocument();
    expect(
      screen.getByText("View todos (login to create or edit todos)"),
    ).toBeInTheDocument();
    expect(screen.getByText("Todo 1")).toBeInTheDocument();
  });

  it("show text for authenticated user", () => {
    const { setUser } = useTodoStore();
    const user = {
      id: "1",
      app_metadata: {},
      user_metadata: {},
      aud: "authenticated",
      email: "test@example.com",
      created_at: "2025-01-01T00:00:00Z",
    };
    setUser(user);
    render(<TodoList />);

    expect(screen.getByText("Todo list")).toBeInTheDocument();
    expect(screen.getByText("Manage your todos")).toBeInTheDocument();
  });
});
