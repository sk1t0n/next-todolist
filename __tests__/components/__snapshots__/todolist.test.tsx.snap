// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`TodoList check snapshot 1`] = `
<DocumentFragment>
  <div
    class="max-w-[600px] mx-auto p-6 space-y-6"
  >
    <div
      class="text-center"
    >
      <h1
        class="text-3xl font-bold mb-2"
      >
        Todo list
      </h1>
      <p
        class="text-muted-foreground"
      >
        View todos (login to create or edit todos)
      </p>
    </div>
    <div
      class="space-y-4"
    >
      <div
        class="rounded-xl border bg-card text-card-foreground shadow"
      >
        <div
          class="p-6"
        >
          <div
            class="flex items-start gap-4"
          >
            <div
              class="flex-1"
            >
              <div
                class="flex items-center gap-2 mb-2"
              >
                <h3
                  class="text-lg font-semibold "
                >
                  Todo 1
                </h3>
              </div>
              <p
                class="text-sm text-muted-foreground"
              >
                Created: 1/1/2025
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
